File: Project/eloteltag/src/main/java/cl/utfsm/elo/eloteltag/ETNube.java
package cl.utfsm.elo.eloteltag;

///import javax.xml.crypto.Data;
import java.awt.geom.Point2D;
import java.util.ArrayList;

/// Clase de simulación del servicio en la nube ETNube, para la localización
/// de equipos
public class ETNube {
    /// Inicializa los datos de la nube como un arreglo complejo de datos tipo Data
    public ETNube() {
        cloudData = new ArrayList<Data>();
    }

    /// Actualiza la localización de un equipo en los datos de la nube
    public void updateLocation(String owner, String equipment, double x, double y) {
        Point2D location;
        if ((location = getLocation(owner, equipment)) == null) {
            location = new Point2D.Double(x, y);
            Data data = new Data(owner, equipment, location);
            cloudData.add(data);
        }
        location.setLocation(x, y);
    }

    /// Obtiene la localización de un equipamiento de acuerdo a los datos en la nube
    public Point2D getLocation(String owner, String equipment) {
        for (Data data : cloudData)
            if (data.ownerName.equals(owner) && data.equipmentName.equals(equipment))
                return data.location;
        return null;
    }

    public String getFindMy(String owner) {
        String h1 = "Bienes de " + owner + "\nÍtems:\n";
        String items = "";
        String h2 = "Dispositivos:\n";
        String devices = "";
        for (Data data : cloudData) {
            if (data.ownerName.equals(owner)) {
                String buffer = (data.equipmentName + ": ("
                        + (int) getLocation(owner, data.equipmentName).getX() + " , "
                        + (int) getLocation(owner, data.equipmentName).getY() + ")\n");
                if (!data.equipmentName.equals("Celular") && !data.equipmentName.equals("Tablet")) {
                    items = items.concat(buffer);
                } else {
                    devices = devices.concat(buffer);
                }
            }
        }
        return h1 + items + h2 + devices;
    }

    /**
     * Obtiene todos los equipos de un dueño específico.
     * Utilizado para GFindMy (extra-crédito) para mostrar
     * las vistas gráficas de los equipos de una persona.
     *
     * @param owner Nombre del dueño
     * @return ArrayList con los datos de los equipos del dueño
     */
    public ArrayList<EquipoInfo> getEquiposByOwner(String owner) {
        ArrayList<EquipoInfo> equipos = new ArrayList<>();
        for (Data data : cloudData) {
            if (data.ownerName.equals(owner)) {
                equipos.add(new EquipoInfo(data.equipmentName, data.location.getX(), data.location.getY()));
            }
        }
        return equipos;
    }

    /**
     * Clase pública para transferir información de equipos.
     * No expone la clase interna Data.
     */
    public static class EquipoInfo {
        public final String Equiponame;
        public final double x;
        public final double y;

        public EquipoInfo(String Equiponame, double x, double y) {
            this.Equiponame = Equiponame;
            this.x = x;
            this.y = y;
        }
    }

    private ArrayList<Data> cloudData;

    /// Define la clase estática interna con la cual se almacenan los datos
    private static class Data {
        public Data(String owner, String equipment, Point2D loc) {
            ownerName = owner;
            equipmentName = equipment;
            location = loc;
        }

        public Point2D location;
        public String ownerName, equipmentName;
    }
}

File: Stage4/etnube.cpp
#include "etnube.h"
ETNube::Data::Data(std::string owner, std::string equipment, QPointF *loc)
{
    this->ownerName = owner;
    this->equipmentName = equipment;
    this->location = loc;
}

ETNube::EquipoInfo::EquipoInfo(std::string equipo_name, double x, double y)
{
    this->x = x;
    this->y = y;
    this->equipo_name = equipo_name;
};

ETNube::ETNube()
{
}
QPointF *ETNube::getLocation(std::string owner, std::string equipment)
{
    for (int i = 0; i < (int)cloudData.size(); i++)
    {
        Data &data = cloudData[i];
        if (data.ownerName == owner && data.equipmentName == equipment)
        {
            return data.location;
        }
    }
    return nullptr;
}

void ETNube::updateLocation(std::string owner, std::string equipment, double x, double y)
{
    QPointF *location = getLocation(owner, equipment);
    if (location == nullptr)
    {
        location = new QPointF(x, y);
        cloudData.push_back(Data(owner, equipment, location));
    }
    location->setX(x);
    location->setY(y);
}
std::vector<ETNube::EquipoInfo> ETNube::getEquiposByOwner(std::string owner)
{
    std::vector<EquipoInfo> equipos;
    for (int i = 0; i < (int)cloudData.size(); i++)
    {
        Data &data = cloudData[i];
        if (data.ownerName == owner)
        {
            equipos.insert(equipos.end(), EquipoInfo(data.equipmentName, data.location->x(), data.location->y()));
        }
    }
    return equipos;
}

std::string ETNube::getFindMy(std::string owner)
{
    std::string h1 = "Bienes de " + owner + "\nÍtems:\n";
    std::string items = "";
    std::string h2 = "Dispositivos:\n";
    std::string devices = "";
    for (int i = 0; i < (int)cloudData.size(); i++)
    {
        Data &data = cloudData[i];
        if (data.ownerName == owner)
        {
            std::string buffer = (data.equipmentName + ": (" + std::to_string((int)getLocation(owner, data.equipmentName)->x()) + " , " + std::to_string((int)getLocation(owner, data.equipmentName)->y()) + ")\n");
            if (data.equipmentName != "Celular" && data.equipmentName != "Tablet")
            {
                items.append(buffer);
            }
            else
            {
                devices.append(buffer);
            }
        }
    }
    return h1 + items + h2 + devices;
}
ETNube::~ETNube()
{
}